server/src/main/java/au/org/aodn/ogcapi/server/common/RestExtService.java

```diff
@@ -0,0 +1,94 @@
+package au.org.aodn.ogcapi.server.common;
+
+import au.org.aodn.ogcapi.server.core.model.CategoryVocabModel;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.RestClientException;
+import org.springframework.web.client.RestTemplate;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Use to store the logic and make it easier to test
+ */
+@Slf4j
+@Service("CommonRestExtService")
+public class RestExtService {
+    @Autowired
+    protected RestTemplate template;
+
+    protected static String path = "/version-2-1/concept.json";
```

Comment:
this PR code looks good to me but I'm confused the purpose of using this particular endpoint to get this response:

the source `url` you send request to is: https://vocabs.ardc.edu.au/repository/api/lda/aodn/aodn-parameter-category-vocabulary/version-2-1/concept.json

after the Json parsing process: the OGC-API endpoint: http://localhost:8082/api/v1/ogc/ext/parameter/categories

returns this kind of response:

```json
[
    {
        "label": "Air pressure",
        "definition": "",
        "broader": {
            "Physical-Atmosphere": "http://vocab.aodn.org.au/def/parameter_classes/category/53"
        },
        "narrower": {}
    },
    {
        "label": "Oxygen",
        "definition": "",
        "broader": {
            "Chemical": "http://vocab.aodn.org.au/def/parameter_classes/category/54"
        },
        "narrower": {}
    },
    {
        "label": "Chlorophyll",
        "definition": "",
        "broader": {
            "Biological": "http://vocab.aodn.org.au/def/parameter_classes/category/55"
        },
        "narrower": {}
    },
    {
        "label": "Wave",
        "definition": "",
        "broader": {
            "Physical-Water": "http://vocab.aodn.org.au/def/parameter_classes/category/56"
        },
        "narrower": {}
    },
    {
        "label": "Visibility",
        "definition": "",
        "broader": {
            "Physical-Atmosphere": "http://vocab.aodn.org.au/def/parameter_classes/category/53"
        },
        "narrower": {}
    },
    {
        "label": "Density",
        "definition": "",
        "broader": {
            "Physical-Water": "http://vocab.aodn.org.au/def/parameter_classes/category/56"
        },
        "narrower": {}
    },
    {
        "label": "Water pressure",
        "definition": "",
        "broader": {
            "Physical-Water": "http://vocab.aodn.org.au/def/parameter_classes/category/56"
        },
        "narrower": {}
    },
    {
        "label": "Other Pigment",
        "definition": "",
        "broader": {
            "Biological": "http://vocab.aodn.org.au/def/parameter_classes/category/55"
        },
        "narrower": {}
    },
    {
        "label": "Alkalinity",
        "definition": "",
        "broader": {
            "Chemical": "http://vocab.aodn.org.au/def/parameter_classes/category/54"
        },
        "narrower": {}
    },
    {
        "label": "Turbidity",
        "definition": "",
        "broader": {
            "Physical-Water": "http://vocab.aodn.org.au/def/parameter_classes/category/56"
        },
        "narrower": {}
    },
    {
        "label": "Current",
        "definition": "",
        "broader": {
            "Physical-Water": "http://vocab.aodn.org.au/def/parameter_classes/category/56"
        },
        "narrower": {}
    },
    {
        "label": "Air-Sea Fluxes",
        "definition": "",
        "broader": {
            "Physical-Water": "http://vocab.aodn.org.au/def/parameter_classes/category/56",
            "Physical-Atmosphere": "http://vocab.aodn.org.au/def/parameter_classes/category/53"
        },
        "narrower": {}
    },
    {
        "label": "Ocean Biota",
        "definition": "",
        "broader": {
            "Biological": "http://vocab.aodn.org.au/def/parameter_classes/category/55"
        },
        "narrower": {}
    },
    {
        "label": "Optical properties",
        "definition": "",
        "broader": {
            "Physical-Water": "http://vocab.aodn.org.au/def/parameter_classes/category/56"
        },
        "narrower": {}
    },
    {
        "label": "Suspended particulate material",
        "definition": "",
        "broader": {
            "Chemical": "http://vocab.aodn.org.au/def/parameter_classes/category/54",
            "Biological": "http://vocab.aodn.org.au/def/parameter_classes/category/55"
        },
        "narrower": {}
    },
    {
        "label": "Backscattering",
        "definition": "",
        "broader": {
            "Physical-Water": "http://vocab.aodn.org.au/def/parameter_classes/category/56"
        },
        "narrower": {}
    },
    {
        "label": "UV radiation",
        "definition": "",
        "broader": {
            "Physical-Atmosphere": "http://vocab.aodn.org.au/def/parameter_classes/category/53"
        },
        "narrower": {}
    },
    {
        "label": "Temperature",
        "definition": "",
        "broader": {
            "Physical-Water": "http://vocab.aodn.org.au/def/parameter_classes/category/56"
        },
        "narrower": {}
    },
    {
        "label": "Acoustics",
        "definition": "",
        "broader": {
            "Physical-Water": "http://vocab.aodn.org.au/def/parameter_classes/category/56"
        },
        "narrower": {}
    },
    {
        "label": "Salinity",
        "definition": "",
        "broader": {
            "Physical-Water": "http://vocab.aodn.org.au/def/parameter_classes/category/56"
        },
        "narrower": {}
    },
    {
        "label": "Nutrient",
        "definition": "",
        "broader": {},
        "narrower": {}
    },
    {
        "label": "Carbon",
        "definition": "",
        "broader": {},
        "narrower": {}
    },
    {
        "label": "Physical-Atmosphere",
        "definition": "",
        "broader": {},
        "narrower": {
            "UV radiation": "http://vocab.aodn.org.au/def/parameter_classes/category/48",
            "Air pressure": "http://vocab.aodn.org.au/def/parameter_classes/category/10",
            "Humidity": "http://vocab.aodn.org.au/def/parameter_classes/category/8",
            "Precipitation and evaporation": "http://vocab.aodn.org.au/def/parameter_classes/category/9",
            "Visibility": "http://vocab.aodn.org.au/def/parameter_classes/category/22",
            "Air-Sea Fluxes": "http://vocab.aodn.org.au/def/parameter_classes/category/30"
        }
    },
    {
        "label": "Chemical",
        "definition": "",
        "broader": {},
        "narrower": {
            "Suspended particulate material": "http://vocab.aodn.org.au/def/parameter_classes/category/46",
            "Oxygen": "http://vocab.aodn.org.au/def/parameter_classes/category/11",
            "Alkalinity": "http://vocab.aodn.org.au/def/parameter_classes/category/27"
        }
    },
    {
        "label": "Biological",
        "definition": "",
        "broader": {},
        "narrower": {
            "Suspended particulate material": "http://vocab.aodn.org.au/def/parameter_classes/category/46",
            "Chlorophyll": "http://vocab.aodn.org.au/def/parameter_classes/category/19",
            "Other Pigment": "http://vocab.aodn.org.au/def/parameter_classes/category/26",
            "Ocean Biota": "http://vocab.aodn.org.au/def/parameter_classes/category/4"
        }
    },
    {
        "label": "Physical-Water",
        "definition": "",
        "broader": {},
        "narrower": {
            "Wave": "http://vocab.aodn.org.au/def/parameter_classes/category/2",
            "Water pressure": "http://vocab.aodn.org.au/def/parameter_classes/category/25",
            "Temperature": "http://vocab.aodn.org.au/def/parameter_classes/category/49",
            "Optical properties": "http://vocab.aodn.org.au/def/parameter_classes/category/45",
            "Backscattering": "http://vocab.aodn.org.au/def/parameter_classes/category/47",
            "Turbidity": "http://vocab.aodn.org.au/def/parameter_classes/category/28",
            "Bathymetry": "http://vocab.aodn.org.au/def/parameter_classes/category/1",
            "Air-Sea Fluxes": "http://vocab.aodn.org.au/def/parameter_classes/category/30",
            "Acoustics": "http://vocab.aodn.org.au/def/parameter_classes/category/5",
            "Density": "http://vocab.aodn.org.au/def/parameter_classes/category/23",
            "Current": "http://vocab.aodn.org.au/def/parameter_classes/category/3",
            "Salinity": "http://vocab.aodn.org.au/def/parameter_classes/category/50"
        }
    },
    {
        "label": "Sea surface height",
        "definition": "",
        "broader": {},
        "narrower": {}
    },
    {
        "label": "Depth",
        "definition": "",
        "broader": {},
        "narrower": {}
    },
    {
        "label": "Wind",
        "definition": "",
        "broader": {},
        "narrower": {}
    },
    {
        "label": "Air temperature",
        "definition": "",
        "broader": {},
        "narrower": {}
    },
    {
        "label": "Humidity",
        "definition": "",
        "broader": {
            "Physical-Atmosphere": "http://vocab.aodn.org.au/def/parameter_classes/category/53"
        },
        "narrower": {}
    },
    {
        "label": "Precipitation and evaporation",
        "definition": "",
        "broader": {
            "Physical-Atmosphere": "http://vocab.aodn.org.au/def/parameter_classes/category/53"
        },
        "narrower": {}
    },
    {
        "label": "Bathymetry",
        "definition": "",
        "broader": {
            "Physical-Water": "http://vocab.aodn.org.au/def/parameter_classes/category/56"
        },
        "narrower": {}
    }
]
```

which can be retrieved by calling direct ARDC API at: https://vocabs.ardc.edu.au/repository/api/lda/aodn/aodn-parameter-category-vocabulary/version-2-1/concept/topConcepts.json

```json
{
   "items":[
      {
         "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/53",
         "dctermsCreated":"2014-06-01T00:00:00Z",
         "dctermsModified":"2018-03-14T22:42:26Z",
         "definition":"This category contains vocabulary terms describing physical atmosphere parameters",
         "narrower":[
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/10",
               "prefLabel":{
                  "_value":"Air pressure",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/22",
               "prefLabel":{
                  "_value":"Visibility",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/30",
               "prefLabel":{
                  "_value":"Air-Sea Fluxes",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/48",
               "prefLabel":{
                  "_value":"UV radiation",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/6",
               "prefLabel":{
                  "_value":"Wind",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/7",
               "prefLabel":{
                  "_value":"Air temperature",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/8",
               "prefLabel":{
                  "_value":"Humidity",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/9",
               "prefLabel":{
                  "_value":"Precipitation and evaporation",
                  "_lang":"en"
               }
            }
         ],
         "prefLabel":{
            "_value":"Physical-Atmosphere",
            "_lang":"en"
         }
      },
      {
         "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/54",
         "dctermsCreated":"2014-06-01T00:00:00Z",
         "dctermsModified":"2022-10-19T22:22:24Z",
         "definition":"This category contains vocabulary terms describing chemicalparameters",
         "narrower":[
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/11",
               "prefLabel":{
                  "_value":"Oxygen",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/27",
               "prefLabel":{
                  "_value":"Alkalinity",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/46",
               "prefLabel":{
                  "_value":"Suspended particulate material",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/51",
               "prefLabel":{
                  "_value":"Nutrient",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/52",
               "prefLabel":{
                  "_value":"Carbon",
                  "_lang":"en"
               }
            }
         ],
         "prefLabel":{
            "_value":"Chemical",
            "_lang":"en"
         }
      },
      {
         "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/55",
         "dctermsCreated":"2014-06-01T00:00:00Z",
         "dctermsModified":"2018-03-14T22:42:27Z",
         "definition":"This category contains vocabulary terms describing biologicalparameters",
         "narrower":[
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/19",
               "prefLabel":{
                  "_value":"Chlorophyll",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/26",
               "prefLabel":{
                  "_value":"Other Pigment",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/4",
               "prefLabel":{
                  "_value":"Ocean Biota",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/46",
               "prefLabel":{
                  "_value":"Suspended particulate material",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/51",
               "prefLabel":{
                  "_value":"Nutrient",
                  "_lang":"en"
               }
            }
         ],
         "prefLabel":{
            "_value":"Biological",
            "_lang":"en"
         }
      },
      {
         "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/56",
         "dctermsCreated":"2014-06-01T00:00:00Z",
         "dctermsModified":"2018-03-14T22:42:28Z",
         "definition":"This category contains vocabulary terms describing physical water parameters",
         "narrower":[
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/1",
               "prefLabel":{
                  "_value":"Bathymetry",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/2",
               "prefLabel":{
                  "_value":"Wave",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/23",
               "prefLabel":{
                  "_value":"Density",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/25",
               "prefLabel":{
                  "_value":"Water pressure",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/28",
               "prefLabel":{
                  "_value":"Turbidity",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/3",
               "prefLabel":{
                  "_value":"Current",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/30",
               "prefLabel":{
                  "_value":"Air-Sea Fluxes",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/45",
               "prefLabel":{
                  "_value":"Optical properties",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/47",
               "prefLabel":{
                  "_value":"Backscattering",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/49",
               "prefLabel":{
                  "_value":"Temperature",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/5",
               "prefLabel":{
                  "_value":"Acoustics",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/50",
               "prefLabel":{
                  "_value":"Salinity",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/57",
               "prefLabel":{
                  "_value":"Sea surface height",
                  "_lang":"en"
               }
            },
            {
               "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/58",
               "prefLabel":{
                  "_value":"Depth",
                  "_lang":"en"
               }
            }
         ],
         "prefLabel":{
            "_value":"Physical-Water",
            "_lang":"en"
         }
      }
   ]
}
```

the structure of the two JSON responses are different but the contents basically identical:

yours is a mixed list of `second-level` and `top-level` vocabs in the same response.

```json
[
  # second-level vocab
  {
      "label": "Air pressure",
      "definition": "",
      "broader": {
          "Physical-Atmosphere": "http://vocab.aodn.org.au/def/parameter_classes/category/53"
      },
      "narrower": {}
  },
  # top-level vocab
  {
      "label": "Physical-Atmosphere",
      "definition": "",
      "broader": {},
      "narrower": {
          "UV radiation": "http://vocab.aodn.org.au/def/parameter_classes/category/48",
          "Air pressure": "http://vocab.aodn.org.au/def/parameter_classes/category/10",
          "Humidity": "http://vocab.aodn.org.au/def/parameter_classes/category/8",
          "Precipitation and evaporation": "http://vocab.aodn.org.au/def/parameter_classes/category/9",
          "Visibility": "http://vocab.aodn.org.au/def/parameter_classes/category/22",
          "Air-Sea Fluxes": "http://vocab.aodn.org.au/def/parameter_classes/category/30"
      }
  },
]
```

ARDC `/topConcepts.json` response is not a mixed list, but list of top-level vocab and their `narrower` (second-level):

```json
{
   "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/53",
   "dctermsCreated":"2014-06-01T00:00:00Z",
   "dctermsModified":"2018-03-14T22:42:26Z",
   "definition":"This category contains vocabulary terms describing physical atmosphere parameters",
   "narrower":[
      {
         "_about":"http://vocab.aodn.org.au/def/parameter_classes/category/10",
         "prefLabel":{
            "_value":"Air pressure",
            "_lang":"en"
         }
      },
     ... other second-level vocab that sitting under `Physical-Atmosphere`
   ],
  "prefLabel": {
    "_value": "Physical-Atmosphere",
    "_lang": "en"
  }
}
```

no need for pagination request e.g `?_page=2`

what will be the difference if `CategoryVocabModel.java` negotiate to map with ARDC response from `/topConcepts.json` endpoint and potentially less code for the `getParameterCategory` method?

Reply:
and I believe this implementation is for the `front-end` to fetch available vocabs; given we just want to display second-level vocabs; if the API returns mixed list; have to implement conditional check to identify which one is `top-level` to ignore from being displayed? then having mixed response can be unnecessary?

Reply:
The reason is because I want to cache the result in ogcapi instead of having the webapp visit it every-time, so to speed up the load as well as avoid hitting the ARDC endpoints too much.

Reply:
Now I understand what you mean, yes using topConcepts.json makes more sense then looping and able to get to the second level, what I am not sure if how we can get to the third level?

Reply:
@vietnguyengit updated now it will consider the third level